ompi/mca/osc/ucx/osc_ucx_passive_target.c

```diff
@@ -0,0 +1,374 @@
+    ompi_osc_ucx_module_t *module = (ompi_osc_ucx_module_t *)win->w_osc_module;
+    ompi_osc_ucx_lock_t *lock = NULL, *item = NULL;
+    ucs_status_t status;
+    int ret = OMPI_SUCCESS;
+    ucp_ep_h ep;
+
+    if (module->epoch_type.access != PASSIVE_EPOCH) {
+        return OMPI_ERR_RMA_SYNC;
+    }
+
+    OPAL_LIST_FOREACH(item, &module->outstanding_locks, ompi_osc_ucx_lock_t) {
+        if (item->target_rank == target) {
+            lock = item;
+            break;
+        }
+    }
+
+    if (lock == NULL) {
+        return OMPI_ERR_RMA_SYNC;
+    }
+
+    opal_list_remove_item(&module->outstanding_locks, &lock->super);
+
+    ep = OSC_UCX_GET_EP(module->comm, target);
+    status = ucp_ep_flush(ep);
+    if (status != UCS_OK) {
+        return OMPI_ERROR;
+    }
+
+    module->global_ops_num -= module->per_target_ops_nums[target];
+    module->per_target_ops_nums[target] = 0;
+
+    if (lock->is_nocheck == false) {
+        if (lock->type == LOCK_EXCLUSIVE) {
+            ret = end_exclusive(module, target);
+        } else {
+            ret = end_shared(module, target);
+        }
+    }
+
+    OBJ_RELEASE(lock);
+
+    module->lock_count--;
+    assert(module->lock_count >= 0);
+    if (module->lock_count == 0) {
+        module->epoch_type.access = NONE_EPOCH;
+        assert(module->global_ops_num == 0);
+    }
+
+    return ret;
+}
+
+int ompi_osc_ucx_lock_all(int assert, struct ompi_win_t *win) {
+    ompi_osc_ucx_module_t *module = (ompi_osc_ucx_module_t*) win->w_osc_module;
+    ompi_osc_ucx_lock_t* lock;
+    int ret = OMPI_SUCCESS;
+
+    if (module->epoch_type.access != NONE_EPOCH &&
+        module->epoch_type.access != FENCE_EPOCH) {
+        return OMPI_ERR_RMA_SYNC;
+    }
+
+    module->epoch_type.access = PASSIVE_ALL_EPOCH;
+
+    assert(module->lock_count == 0);
+
+    lock = OBJ_NEW(ompi_osc_ucx_lock_t);
+    lock->target_rank = -1;
+    lock->type = LOCK_SHARED;
+
+    if (0 == (assert & MPI_MODE_NOCHECK)) {
+        int i, comm_size;
+        lock->is_nocheck = false;
+        comm_size = ompi_comm_size(module->comm);
+        for (i = 0; i < comm_size; i++) {
+            ret |= start_shared(module, i);
+        }
+    } else {
+        lock->is_nocheck = true;
+    }
+
+    if (ret == OMPI_SUCCESS) {
+        opal_list_append(&module->outstanding_locks, &lock->super);
+    } else {
+        OBJ_RELEASE(lock);
+    }
+
+    return ret;
+}
+
+int ompi_osc_ucx_unlock_all(struct ompi_win_t *win) {
+    ompi_osc_ucx_module_t *module = (ompi_osc_ucx_module_t*)win->w_osc_module;
+    ompi_osc_ucx_lock_t *lock = NULL, *item = NULL;
+    int comm_size = ompi_comm_size(module->comm);
+    ucs_status_t status;
+    int ret = OMPI_SUCCESS;
+
+    if (module->epoch_type.access != PASSIVE_ALL_EPOCH) {
+        return OMPI_ERR_RMA_SYNC;
+    }
+
+    assert(module->lock_count == 0);
+
+    OPAL_LIST_FOREACH(item, &module->outstanding_locks, ompi_osc_ucx_lock_t) {
+        if (item->target_rank == -1) {
```

Comment:
it can be multiple elements in the list. One rank is allowed to issue lock request to multiple different target ranks, as long as it does not issue more than one request to the same target rank. For example, the follow case is allowed: 

rank0----rank1----rank2
lock(1)
lock(2)
unlock(1)
unlock(2)

In this case, there are two elements in rank0's list.